[2025-03-09T23:39:50.136+0600] {local_task_job_runner.py:123} INFO - ::group::Pre task execution logs
[2025-03-09T23:39:50.149+0600] {taskinstance.py:2614} INFO - Dependencies all met for dep_context=non-requeueable deps ti=<TaskInstance: real_time_e2e_dag.extract_data_task manual__2025-03-09T17:39:27.947870+00:00 [queued]>
[2025-03-09T23:39:50.157+0600] {taskinstance.py:2614} INFO - Dependencies all met for dep_context=requeueable deps ti=<TaskInstance: real_time_e2e_dag.extract_data_task manual__2025-03-09T17:39:27.947870+00:00 [queued]>
[2025-03-09T23:39:50.158+0600] {taskinstance.py:2867} INFO - Starting attempt 1 of 1
[2025-03-09T23:39:50.279+0600] {taskinstance.py:2890} INFO - Executing <Task(PythonOperator): extract_data_task> on 2025-03-09 17:39:27.947870+00:00
[2025-03-09T23:39:50.288+0600] {clientserver.py:543} INFO - Closing down clientserver connection
[2025-03-09T23:39:50.293+0600] {standard_task_runner.py:104} INFO - Running: ['airflow', 'tasks', 'run', 'real_time_e2e_dag', 'extract_data_task', 'manual__2025-03-09T17:39:27.947870+00:00', '--job-id', '14', '--raw', '--subdir', 'DAGS_FOLDER/RT_temp.py', '--cfg-path', '/tmp/tmpgijxqbfd']
[2025-03-09T23:39:50.294+0600] {standard_task_runner.py:105} INFO - Job 14: Subtask extract_data_task
[2025-03-09T23:39:50.298+0600] {logging_mixin.py:190} WARNING - /home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/task/task_runner/standard_task_runner.py:70 DeprecationWarning: This process (pid=31608) is multi-threaded, use of fork() may lead to deadlocks in the child.
[2025-03-09T23:39:50.299+0600] {standard_task_runner.py:72} INFO - Started process 31795 to run task
[2025-03-09T23:39:50.336+0600] {task_command.py:467} INFO - Running <TaskInstance: real_time_e2e_dag.extract_data_task manual__2025-03-09T17:39:27.947870+00:00 [running]> on host shamim-X541UV
[2025-03-09T23:39:50.403+0600] {taskinstance.py:3134} INFO - Exporting env vars: AIRFLOW_CTX_DAG_OWNER='airflow' AIRFLOW_CTX_DAG_ID='real_time_e2e_dag' AIRFLOW_CTX_TASK_ID='extract_data_task' AIRFLOW_CTX_EXECUTION_DATE='2025-03-09T17:39:27.947870+00:00' AIRFLOW_CTX_TRY_NUMBER='1' AIRFLOW_CTX_DAG_RUN_ID='manual__2025-03-09T17:39:27.947870+00:00'
[2025-03-09T23:39:50.405+0600] {logging_mixin.py:190} INFO - Task instance is in running state
[2025-03-09T23:39:50.405+0600] {logging_mixin.py:190} INFO -  Previous state of the Task instance: queued
[2025-03-09T23:39:50.406+0600] {logging_mixin.py:190} INFO - Current task name:extract_data_task state:running start_date:2025-03-09 17:39:50.150279+00:00
[2025-03-09T23:39:50.406+0600] {logging_mixin.py:190} INFO - Dag name:real_time_e2e_dag and current dag run status:running
[2025-03-09T23:39:50.406+0600] {taskinstance.py:732} INFO - ::endgroup::
[2025-03-09T23:39:54.372+0600] {xcom.py:690} ERROR - Object of type DataFrame is not JSON serializable. If you are using pickle instead of JSON for XCom, then you need to enable pickle support for XCom in your airflow config or make sure to decorate your object with attr.
[2025-03-09T23:39:54.375+0600] {taskinstance.py:3313} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/json.py", line 92, in default
    return serialize(o)
           ^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/serialization/serde.py", line 190, in serialize
    raise TypeError(f"cannot serialize object of type {cls}")
TypeError: cannot serialize object of type <class 'pyspark.sql.dataframe.DataFrame'>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 768, in _execute_task
    result = _execute_callable(context=context, **execute_callable_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 734, in _execute_callable
    return ExecutionCallableRunner(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/operator_helpers.py", line 252, in run
    return self.func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/baseoperator.py", line 424, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/operators/python.py", line 238, in execute
    return_value = self.execute_callable()
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/operators/python.py", line 256, in execute_callable
    return runner.run(*self.op_args, **self.op_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/operator_helpers.py", line 252, in run
    return self.func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/Real_Time_Big_Data_Airflow/dags/RT_temp.py", line 47, in extract_data
    kwargs['ti'].xcom_push(key='DataFrame', value=parsed_df)
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/session.py", line 97, in wrapper
    return func(*args, session=session, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 3645, in xcom_push
    XCom.set(
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/api_internal/internal_api_call.py", line 166, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/session.py", line 94, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/xcom.py", line 249, in set
    value = cls.serialize_value(
            ^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/xcom.py", line 688, in serialize_value
    return json.dumps(value, cls=XComEncoder).encode("UTF-8")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
          ^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/json.py", line 105, in encode
    return super().encode(o)
           ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/json.py", line 94, in default
    return super().default(o)
           ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type DataFrame is not JSON serializable
[2025-03-09T23:39:54.393+0600] {logging_mixin.py:190} INFO - Task instance in failure state
[2025-03-09T23:39:54.394+0600] {logging_mixin.py:190} INFO - Task start:2025-03-09 17:39:50.150279+00:00 end:2025-03-09 17:39:54.391973+00:00 duration:4.241694
[2025-03-09T23:39:54.394+0600] {logging_mixin.py:190} INFO - Task:<Task(PythonOperator): extract_data_task> dag:<DAG: real_time_e2e_dag> dagrun:<DagRun real_time_e2e_dag @ 2025-03-09 17:39:27.947870+00:00: manual__2025-03-09T17:39:27.947870+00:00, state:running, queued_at: 2025-03-09 17:39:27.977648+00:00. externally triggered: True>
[2025-03-09T23:39:54.395+0600] {logging_mixin.py:190} INFO - Failure caused by Object of type DataFrame is not JSON serializable
[2025-03-09T23:39:54.395+0600] {taskinstance.py:1226} INFO - Marking task as FAILED. dag_id=real_time_e2e_dag, task_id=extract_data_task, run_id=manual__2025-03-09T17:39:27.947870+00:00, execution_date=20250309T173927, start_date=20250309T173950, end_date=20250309T173954
[2025-03-09T23:39:54.443+0600] {taskinstance.py:341} INFO - ::group::Post task execution logs
[2025-03-09T23:39:54.445+0600] {standard_task_runner.py:124} ERROR - Failed to execute job 14 for task extract_data_task (Object of type DataFrame is not JSON serializable; 31795)
Traceback (most recent call last):
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/json.py", line 92, in default
    return serialize(o)
           ^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/serialization/serde.py", line 190, in serialize
    raise TypeError(f"cannot serialize object of type {cls}")
TypeError: cannot serialize object of type <class 'pyspark.sql.dataframe.DataFrame'>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/task/task_runner/standard_task_runner.py", line 117, in _start_by_fork
    ret = args.func(args, dag=self.dag)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/cli/cli_config.py", line 49, in command
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/cli.py", line 116, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/cli/commands/task_command.py", line 483, in task_run
    task_return_code = _run_task_by_selected_method(args, _dag, ti)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/cli/commands/task_command.py", line 256, in _run_task_by_selected_method
    return _run_raw_task(args, ti)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/cli/commands/task_command.py", line 341, in _run_raw_task
    return ti._run_raw_task(
           ^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/session.py", line 97, in wrapper
    return func(*args, session=session, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 3006, in _run_raw_task
    return _run_raw_task(
           ^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 274, in _run_raw_task
    TaskInstance._execute_task_with_callbacks(
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 3161, in _execute_task_with_callbacks
    result = self._execute_task(context, task_orig)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 3185, in _execute_task
    return _execute_task(self, context, task_orig)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 768, in _execute_task
    result = _execute_callable(context=context, **execute_callable_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 734, in _execute_callable
    return ExecutionCallableRunner(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/operator_helpers.py", line 252, in run
    return self.func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/baseoperator.py", line 424, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/operators/python.py", line 238, in execute
    return_value = self.execute_callable()
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/operators/python.py", line 256, in execute_callable
    return runner.run(*self.op_args, **self.op_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/operator_helpers.py", line 252, in run
    return self.func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/Real_Time_Big_Data_Airflow/dags/RT_temp.py", line 47, in extract_data
    kwargs['ti'].xcom_push(key='DataFrame', value=parsed_df)
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/session.py", line 97, in wrapper
    return func(*args, session=session, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/taskinstance.py", line 3645, in xcom_push
    XCom.set(
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/api_internal/internal_api_call.py", line 166, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/session.py", line 94, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/xcom.py", line 249, in set
    value = cls.serialize_value(
            ^^^^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/models/xcom.py", line 688, in serialize_value
    return json.dumps(value, cls=XComEncoder).encode("UTF-8")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
          ^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/json.py", line 105, in encode
    return super().encode(o)
           ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/home/shamim/airflow_venv/lib/python3.12/site-packages/airflow/utils/json.py", line 94, in default
    return super().default(o)
           ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type DataFrame is not JSON serializable
[2025-03-09T23:39:54.477+0600] {local_task_job_runner.py:266} INFO - Task exited with return code 1
[2025-03-09T23:39:54.535+0600] {taskinstance.py:3901} INFO - 0 downstream tasks scheduled from follow-on schedule check
[2025-03-09T23:39:54.549+0600] {local_task_job_runner.py:245} INFO - ::endgroup::
